### PR TITLE
Add UDS read/write error handling and tests with glide core mock.

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
Features:
1. Error handling on UDS write
2. Error handling on UDS read
3. Glide core mock
4. Test superclass with glide core mock
5. Tests for error handling

Missing:
1. Error handling on channel creation. Left TODO there. Requires significant rework of `ChannelHandler`.
2. IT. Can't be reproduced in IT unfortunately.

### CODE REVIEW

File: java/client/src/main/java/glide/connectors/handlers/ChannelHandler.java [hunk 1]
```diff
@@ -82,4 +87,18 @@ public ChannelFuture close() {
         callbackDispatcher.shutdownGracefully();
         return channel.close();
     }
+
+    /** Propagate an error from Netty's future into the future returned to user. */
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/connectors/handlers/ChannelHandler.java:91]:
```suggestion
    /** Propagate an error from Netty's {@link ChannelFuture} and complete the {@link CompletableFuture} promise. */
```

File: java/client/src/main/java/glide/connectors/resources/ThreadPoolResource.java [hunk 2]
```diff
@@ -12,8 +12,8 @@
  */
 @Getter
 public abstract class ThreadPoolResource {
-    private EventLoopGroup eventLoopGroup;
-    private Class<? extends DomainSocketChannel> domainSocketChannelClass;
+    private final EventLoopGroup eventLoopGroup;
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/connectors/resources/ThreadPoolResource.java:15]: these changes are unrelated, no?

File: java/client/src/main/java/glide/connectors/handlers/ChannelHandler.java [hunk 3]
```diff
@@ -82,4 +87,18 @@ public ChannelFuture close() {
...
+    @RequiredArgsConstructor
+    private static class NettyFutureErrorHandler implements ChannelFutureListener {
+
+        private final CompletableFuture<Response> promise;
+
+        @Override
+        public void operationComplete(@NonNull ChannelFuture channelFuture) throws Exception {
+            if (!channelFuture.isSuccess()) {
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/connectors/handlers/ChannelHandler.java:99]: According to https://netty.io/4.0/api/io/netty/channel/ChannelFuture.html, !isSuccess just means that the future did not complete successfully.  
We should check for `cause != null` to determine if it failed.  
In the case of `isCancelled`, do we feel it is necessary to terminate the client?

File: java/client/src/main/java/glide/connectors/handlers/ChannelHandler.java (hunk 1 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/connectors/handlers/ChannelHandler.java]: Updated in 5b707488

File: java/client/src/main/java/glide/connectors/resources/ThreadPoolResource.java (hunk 2 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/connectors/resources/ThreadPoolResource.java]: Reverted in 5b707488

File: java/client/src/main/java/glide/connectors/handlers/ChannelHandler.java (hunk 3 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/connectors/handlers/ChannelHandler.java:106]: I added extra checks, but I have no idea how to test them. Any clue?

File: java/client/src/test/java/glide/connection/ConnectionWithGlideMockTests.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+import java.util.concurrent.TimeoutException;
+import lombok.SneakyThrows;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import redis_request.RedisRequestOuterClass.RedisRequest;
+import response.ResponseOuterClass.Response;
+
+public class ConnectionWithGlideMockTests extends RustCoreLibMockTestBase {
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/connection/ConnectionWithGlideMockTests.java:33]: Can we add tests for InterruptedExceptions on the thread?

Comment (Yury-Fridlyand) [reply java/client/src/test/java/glide/connection/ConnectionWithGlideMockTests.java:33]: What should I do to get one? Kill a thread?

File: java/client/src/test/java/glide/connection/ConnectionWithGlideMockTests.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        channelHandler.close();
+    }
+
+    private Future<Response> testConnection() {
+        return channelHandler.connect(createConnectionRequest());
+    }
+
+    private static ConnectionRequest createConnectionRequest() {
+        return ConnectionRequest.newBuilder()
+                .addAddresses(NodeAddress.newBuilder().setHost("dummyhost").setPort(42).build())
```
Comment (jduo) [inline java/client/src/test/java/glide/connection/ConnectionWithGlideMockTests.java:56]: Let's add a comment about why fake values are used.

File: java/client/src/test/java/glide/connection/ConnectionWithGlideMockTests.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        stopRustCoreLibMock();
+        try {
+            var exception =
+                    assertThrows(
+                            ExecutionException.class, () -> client.customCommand(new String[0]).get(1, SECONDS));
+            assertTrue(exception.getCause() instanceof RuntimeException);
+
+            // Not a public class, can't import
+            assertEquals(
+                    "io.netty.channel.StacklessClosedChannelException",
```
Comment (jduo) [inline java/client/src/test/java/glide/connection/ConnectionWithGlideMockTests.java:189]: This seems brittle if Netty's version changes.

File: java/client/src/test/java/glide/utils/RustCoreMock.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import lombok.SneakyThrows;
+import redis_request.RedisRequestOuterClass.RedisRequest;
+import response.ResponseOuterClass.ConstantResponse;
+import response.ResponseOuterClass.Response;
+
+public class RustCoreMock {
+
+    public abstract static class GlideMock {
```
Comment (jduo) [inline java/client/src/test/java/glide/utils/RustCoreMock.java:31]: This could be an interface with a default isRaw() function (as well as be made into a @FunctionalInterface) to make this a bit more flexible.

File: java/client/src/test/java/glide/utils/RustCoreMock.java [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+                                                            @Override
+                                                            public void channelRead(ChannelHandlerContext ctx, Object msg)
+                                                                    throws Exception {
+                                                                var buf = (ByteBuf) msg;
+                                                                var bytes = new byte[buf.readableBytes()];
+                                                                buf.readBytes(bytes);
+                                                                buf.release();
+                                                                if (messageProcessor.isRaw()) {
+                                                                    ch.writeAndFlush(
+                                                                            Unpooled.copiedBuffer(messageProcessor.handle(bytes)));
```
Comment (jduo) [inline java/client/src/test/java/glide/utils/RustCoreMock.java:137]: This has alot of nesting. Anyway it can be reduced? It's deep enough that it's pretty unreadable in GitHub's diff view.

File: java/client/src/test/java/glide/connection/ConnectionWithGlideMockTests.java (hunk 4 above)
Comment (acarbonetto) [reply java/client/src/test/java/glide/connection/ConnectionWithGlideMockTests.java:33]: yes?

File: java/client/src/test/java/glide/connection/ConnectionWithGlideMockTests.java (hunk 6 above)
Comment (acarbonetto) [reply java/client/src/test/java/glide/connection/ConnectionWithGlideMockTests.java:189]: hopefully they make it public when they update?

File: java/client/src/test/java/glide/utils/RustCoreMock.java (hunk 8 above)
Comment (acarbonetto) [reply java/client/src/test/java/glide/utils/RustCoreMock.java:138]: @Yury-Fridlyand at least create a separate class for ChannelInboundHandlerAdapter?

File: java/client/src/test/java/glide/connection/ConnectionWithGlideMockTests.java (hunk 6 above)
Comment (Yury-Fridlyand) [reply java/client/src/test/java/glide/connection/ConnectionWithGlideMockTests.java:182]: I hope it won't happen without CI verification to catch failures.

File: java/client/src/test/java/glide/utils/RustCoreMock.java (hunk 7 above)
Comment (Yury-Fridlyand) [reply java/client/src/test/java/glide/utils/RustCoreMock.java]: Added in f4699ada

**Review by acarbonetto — APPROVED** [review]
lgtm
